src/sentry/spans/buffer_v2.py

```diff
@@ -0,0 +1,290 @@
+#
+# The segment ID in the Kafka protocol is actually only the span ID.
+SegmentId = bytes
+
+
+def get_redis_client() -> RedisCluster[bytes] | StrictRedis[bytes]:
+    return redis.redis_clusters.get_binary(settings.SENTRY_SPAN_BUFFER_CLUSTER)
+
+
+add_buffer_script = redis.load_redis_script("spans/add-buffer.lua")
+
+
+def segment_to_span_id(segment_id: SegmentId) -> bytes:
+    return segment_id.rsplit(b":", 1)[1]
+
+
+# fun fact: namedtuples are faster to construct than dataclasses
+class Span(NamedTuple):
+    trace_id: str
+    span_id: str
+    parent_span_id: str | None
+    project_id: int
+    payload: bytes
+    is_segment_span: bool = False
+
+
+class RedisSpansBufferV2:
+    def __init__(
+        self,
+        assigned_shards: list[int],
+        span_buffer_timeout_secs: int = 60,
+        span_buffer_root_timeout_secs: int = 10,
+        redis_ttl: int = 3600,
+    ):
+        self.client: RedisCluster[bytes] | StrictRedis[bytes] = get_redis_client()
+        self.assigned_shards = list(assigned_shards)
+        self.span_buffer_timeout_secs = span_buffer_timeout_secs
+        self.span_buffer_root_timeout_secs = span_buffer_root_timeout_secs
+        self.redis_ttl = redis_ttl
+
+    # make it pickleable
+    def __reduce__(self):
+        return (
+            RedisSpansBufferV2,
+            (
+                self.assigned_shards,
+                self.span_buffer_timeout_secs,
+                self.span_buffer_root_timeout_secs,
+                self.redis_ttl,
+            ),
+        )
+
+    def _is_root_span(self, span: Span) -> bool:
+        return span.parent_span_id is None or span.is_segment_span
+
+    def process_spans(self, spans: Sequence[Span], now: int):
+        queue_keys = []
+        queue_delete_items = []
+        queue_items = []
+        queue_item_has_root_span = []
+
+        is_root_span_count = 0
+        has_root_span_count = 0
+
+        with metrics.timer("sentry.spans.buffer.process_spans.step1"):
+            with self.client.pipeline(transaction=False) as p:
+                for span in spans:
+                    # (parent_span_id) -> [Span]
+                    shard = self.assigned_shards[int(span.trace_id, 16) % len(self.assigned_shards)]
+                    queue_key = f"span-buf:q:{shard}"
+                    parent_span_id = span.parent_span_id or span.span_id
+
+                    is_root_span = self._is_root_span(span)
+                    if is_root_span:
+                        is_root_span_count += 1
+
+                    # hack to make redis-cluster-py pipelines work well with
+                    # scripts. we cannot use EVALSHA or "the normal way to do
+                    # scripts in sentry" easily until we get rid of
+                    # redis-cluster-py sentrywide. this probably leaves a bit of
+                    # perf on the table as we send the full lua sourcecode with every span.
+                    p.eval(
+                        add_buffer_script.script,
```

Comment:
We are highly likely to find spans of the same trace in a batch. This means we could pre-process them in memory and minimize the numbers of mutations in Redis. For example, we should build partial trees in memory, then insert redirect keys for the top-most parent known at the current time.

Reply:
I generally don't believe that reducing the amount of work done in Redis will pay off if we pay the cost for the same CPU-bound work but executed in Python. So while Redis is a scaling bottleneck I do believe it's just more _cost efficient_ to run things like tree construction within Redis, and scaling out Redis to the degree we need it is actually not a problem.

Of course I say this without having it rigorously tested, but I don't think we should spend time on that. If this consumer was Rust then I would've been all-in on the local buffer approach.